### PR TITLE
Custom token response in AuthorizationCodeGrant

### PR DESCRIPTION
**Pull request by majk-p**

### CODE REVIEW

File: oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCode.scala [hunk 1]
```diff
@@ -83,8 +91,8 @@ object AuthorizationCode {
           .body(refreshTokenRequestParams(refreshToken, clientId, clientSecret.value, scopeOverride.toRequestMap))
           .response(asString)
       }
-      .map(_.body.leftMap(new RuntimeException(_)).flatMap(decode[RefreshTokenResponse]).toTry)
-      .map(_.map(_.toOauth2Token(refreshToken)))
+      .map(_.body.leftMap(new RuntimeException(_)).flatMap(decode[ResponseType]).toTry)
+      // .map(_.map(_.toOauth2Token(refreshToken)))
```
Comment (tplaskowski) [inline oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCode.scala:95]: To remove?

File: oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCodeProvider.scala [hunk 2]
```diff
@@ -6,15 +6,18 @@ import eu.timepit.refined.refineV
...
   *  Use the `instance` companion object method to create instances.
   *
   * @tparam UriType type of returned uri. Supported types are:
   *                 Refined[String, Url] and Uri
+  * @tparam TokenType type that models token response. It must implement MinimalStructurem, and have io.circe.Decoder instance.
+  *                   Predefined implementations: OAuth2TokenResponse and ExtendedOAuth2TokenResponse
   * @tparam F effect wrapper
   */
-trait AuthorizationCodeProvider[UriType, F[_]] {
+trait AuthorizationCodeProvider[UriType, TokenType, F[_]] {
```
Comment (tplaskowski) [inline oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCodeProvider.scala:20]: In all places. I would change `TokenType`/`ResponseType` to just `Response` or use a letter or two (like `TT`).

File: oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/Oauth2TokenResponse.scala [hunk 3]
```diff
@@ -4,7 +4,61 @@ import io.circe.Decoder
...
+    *      OPTIONAL.  The refresh token, which can be used to obtain new
+    *      access tokens using the same authorization grant as described
+    *      in Section 6.
+    *
+    *scope
+    *      OPTIONAL, if identical to the scope requested by the client;
+    *      otherwise, REQUIRED.  The scope of the access token as
+    *      described by Section 3.3.
+    */
+  trait MinimalStructure {
```
Comment (tplaskowski) [inline oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/Oauth2TokenResponse.scala:44]: Maybe `Basic` or `Base`? You are referencing it using object so `OAuth2TokenResponse.MinimalStructure` which is kind of lengthy, `OAuth2TokenResponse.Basic` seems more readable.

**Review by tplaskowski — APPROVED** [review]
All comments are rather minor 🙂

File: oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/Oauth2TokenResponse.scala (hunk 3 above)
Comment (majk-p) [reply oauth2/src/main/scala/com/ocadotechnology/sttp/oauth2/Oauth2TokenResponse.scala]: `Base` sounds good, changed

File: oauth2/src/test/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCodeSpec.scala [hunk 4]
```diff
@@ -5,6 +5,11 @@ import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpec
 import sttp.model.Uri
 import AuthorizationCodeProvider.Config._
+import sttp.client3.testing._
+import scala.util.Try
+import sttp.monad.TryMonad
+// import scala.util.Failure
```
[the remark this answers is not in the document]
Comment (majk-p) [reply oauth2/src/test/scala/com/ocadotechnology/sttp/oauth2/AuthorizationCodeSpec.scala]: yup, done

**Comment by majk-p** [comment]
Guess this partially addresses https://github.com/ocadotechnology/sttp-oauth2/issues/11

**Review by kubukoz — APPROVED** [review]

**Comment by majk-p** [comment]
Accepting this MR with broken pipeline, as it's only a failed MiMa check. It will be fixed automatically when tagged as `v0.10.0`.

**Pull request merged**